packages/form-js-viewer/src/features/expressionLanguage/ConditionChecker.js

```diff
@@ -17,37 +17,108 @@ export default class ConditionChecker {
    *
-   * @param {Object<string, any>} properties
    * @param {Object<string, any>} data
+   * @param {Object<string, any>} contextData
    * @param {Object} [options]
    * @param {Function} [options.getFilterPath]
+   * @param {boolean} [options.leafNodeDeletionOnly]
    */
-  applyConditions(properties, data = {}, options = {}) {
+  applyConditions(data, contextData = {}, options = {}) {
 
-    const newProperties = clone(properties);
+    const workingData = clone(data);
 
     const {
-      getFilterPath = (field) => this._pathRegistry.getValuePath(field)
+      getFilterPath = (field, indexes) => this._pathRegistry.getValuePath(field, { indexes }),
+      leafNodeDeletionOnly = false
     } = options;
 
+    const _applyConditionsWithinScope = (rootField, scopeContext, startHidden = false) => {
+
+      const {
+        indexes = {},
+        expressionIndexes = [],
+        scopeData = contextData,
+        parentScopeData = null
+      } = scopeContext;
+
+      this._pathRegistry.executeRecursivelyOnFields(rootField, ({ field, isClosed, isRepeatable, context }) => {
+
+        const {
+          conditional,
+          components,
+          id
+        } = field;
+
+        // build the expression context in the right format
+        const localExpressionContext = buildExpressionContext({
+          this: scopeData,
+          data: contextData,
+          i: expressionIndexes,
+          parent: parentScopeData
+        });
+
+        context.isHidden = startHidden || context.isHidden || (conditional && this._checkHideCondition(conditional, localExpressionContext));
+
+        // if a field is repeatable and visible, we need to implement custom recursion on its children
+        if (isRepeatable && (!context.isHidden || leafNodeDeletionOnly)) {
+
+          // prevent the regular recursion behavior of executeRecursivelyOnFields
+          context.preventRecursion = true;
+
+          const repeaterValuePath = this._pathRegistry.getValuePath(field, { indexes });
+          const repeaterValue = get(contextData, repeaterValuePath);
+
+          // quit early if there are no children or data associated with the repeater
+          if (!Array.isArray(repeaterValue) || !repeaterValue.length || !Array.isArray(components) || !components.length) {
+            return;
+          }
+
+          for (let i = 0; i < repeaterValue.length; i++) {
+
+            // create a new scope context for each index
+            const newScopeContext = {
+              indexes: { ...indexes, [id]: i },
+              expressionIndexes: [ ...expressionIndexes, i + 1 ],
+              scopeData: repeaterValue[i],
+              parentScopeData: scopeData
+            };
+
+            // for each child component, apply conditions within the new repetition scope
+            components.forEach(component => {
+              _applyConditionsWithinScope(component, newScopeContext, context.isHidden);
+            });
```

Comment:
This looks a dangerous, a loop inside of a loop 😬

Reply:
Yup that's the brute forcing bit I mentioned on call. I want to get rid of this class, have an issue https://github.com/bpmn-io/form-js/issues/924 for a refactor that I plan to tackle later. The way we do hide-if at the moment just isn't very scalable which kinda forced me into building this spaghetti.

Reply:
But it's literally what we want, a loop inside of a loop. Because we have a list of children repeated a certain amount of time. Whether we unwrap it or not this will be the nature of this code. There's just nicer ways to do this that involve some restructuring.